src/nccl_ofi_rdma.c

```diff
@@ -5204,6 +5204,15 @@ static int connect(nccl_net_ofi_ep_t *base_ep,
 
 static void ep_rail_release(nccl_net_ofi_ep_rail_t *rail, int dev_id)
 {
+	if (ofi_nccl_endpoint_per_communicator() != 0) {
```

Comment:
Something seems wrong here. We always create the device-cq, but we only avoid freeing it in `ep_rail_release` if `ofi_nccl_endpoint_per_communicator()`. In that case, the `endpoint_per_comm == 0` case is probably broken: we can free the device-cq here and then not allocate it again.

Instead, we should probably only allocate the device-shared CQ if `endpoint_per_comm != 0`.

Reply:
Actually, for simplicity, I think we should just always allocate the device-shared CQ, and never free it in `ep_rail_release`. We can free it in the device's finalize method, which will never be called by NCCL today.

Reply:
Fixed in next rev

Reply:
I'm fine with your second update.  I do object to allocating "device resources", which includes CQs, unless we're going to use them.